packages/portfolio-contract/tools/plan-solve.ts

```diff
@@ -388,72 +406,145 @@ export const solveRebalance = async (
   return { flows, detail: { solution } };
 };
 
-export const rebalanceMinCostFlowSteps = async (
+/**
+ * Compute partial order from solved flows and initial supplies.
+ *
+ * For each node, we track available supply (initial + completed inflows).
+ * An outflow step depends on the minimum set of inflow steps needed to
+ * provide sufficient supply.
+ *
+ * This approach correctly handles cases like:
+ * - Multiple operations to/from the same node with separate dependencies
+ * - Initial supplies that satisfy some outflows immediately
+ * - Fan-out patterns where multiple outflows depend on the same inflow
+ *
+ * @param flows - solved flows from the LP solver
+ * @param initialSupplies - starting balances at each node
+ * @returns prioritized flows in execution order and partial order constraints
+ */
+export const computePartialOrder = (
   flows: SolvedEdgeFlow[],
-  graph: FlowGraph,
-  {
-    brand,
-    feeBrand,
-    gasEstimator,
-  }: {
-    brand: NatAmount['brand'];
-    feeBrand: NatAmount['brand'];
-    gasEstimator: GasEstimator;
-  },
-): Promise<MovementDesc[]> => {
-  const supplies = new Map(
-    typedEntries(graph.supplies).filter(([_place, amount]) => amount > 0),
+  initialSupplies: Map<AssetPlaceRef, number>,
+): { prioritized: SolvedEdgeFlow[]; order?: StepOrder } => {
+  // Return immediately if there's nothing to do.
+  if (flows.length === 0) return { prioritized: [] };
+
+  type FlowIndex = number;
+  type Inflow = { stepIdx?: number; unclaimed: number };
+
+  /** Instantaneous supply by node */
+  const available = new Map(initialSupplies);
+
+  /** Step indices constituting those supplies (starting with just the initial supply) */
+  const inflows = new Map<AssetPlaceRef, Inflow[]>(
+    [...initialSupplies.entries()].map(([place, value]) => {
+      const nullInflow = { stepIdx: undefined, unclaimed: value };
+      return [place, [nullInflow]];
+    }),
   );
 
-  type AnnotatedFlow = SolvedEdgeFlow & { srcChain: string };
-  const pendingFlows = new Map<string, AnnotatedFlow>(
-    flows
-      .filter(f => f.flow > FLOW_EPS)
-      .map(f => [f.edge.id, { ...f, srcChain: chainOf(f.edge.src) }]),
-  );
-  const prioritized = [] as AnnotatedFlow[];
+  const scheduled = new Set<FlowIndex>();
+  const order: StepOrder = [];
 
   // Maintain last chosen originating chain to group sequential operations.
   let lastChain: string | undefined;
 
-  while (pendingFlows.size) {
-    // Find flows that can be executed based on current supplies.
-    const candidates = [...pendingFlows.values()].filter(
-      f => (supplies.get(f.edge.src) || 0) >= f.flow,
-    );
+  while (scheduled.size < flows.length) {
+    /** Flows that can execute with current available supply */
+    const candidates: { flowIdx: FlowIndex; flow: SolvedEdgeFlow }[] = [];
+    for (let i = 0; i < flows.length; i += 1) {
+      if (scheduled.has(i)) continue;
+      const flow = flows[i];
+      const srcSupply = available.get(flow.edge.src) || 0;
```

Comment:
Isn't 0 a valid flowID?  If so, could there be an issue with the `|| 0` returned index?  Could we end up adding "step 0 relies on step 0"?